### PR TITLE
Support validated links coming from e.com

### PR DESCRIPTION
**Pull request by iwiznia**
Please review!

CC: @cead22, @tgolen, @marcaaron, @Luke9389 
### Fixed Issues
https://github.com/Expensify/Expensify/issues/177739

### Tests
See https://github.com/Expensify/Web-Expensify/pull/31977

### QA Steps
1. Log into an account on NewDot Expensify, create a Workspace if you don't already have one using the `+` on the bottom right.
2. Log out of NewDot and then log into that same account on OldDot and go to Settings -> Policy and find your Workspace there. It should have a card icon like these:
image
3. Click on its name and verify it opens a new tab, automatically logs you in, and brings you to the workspace settings page:
image

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
Included above

### CODE REVIEW

**Comment by TomatoToaster** [comment]
PR for adding workspace/new is ready here: https://github.com/Expensify/App/pull/5396/files. This could be its own PR and I could add testing steps for it.

Or I could merge it with this one which might help test that this PR works. I'm also trying to make sure urls where we do `exitTo=workspace/<workspaceID>/card` also works. 

I'm going to work on the Web-Expensify part right now.

File: src/libs/Navigation/linkingConfig.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ export default {
             [SCREENS.LOGIN_WITH_VALIDATE_CODE_2FA_NEW_WORKSPACE]: ROUTES.LOGIN_WITH_VALIDATE_CODE_2FA_NEW_WORKSPACE,
             [SCREENS.LOGIN_WITH_VALIDATE_CODE_WORKSPACE_CARD]: ROUTES.LOGIN_WITH_VALIDATE_CODE_WORKSPACE_CARD,
             [SCREENS.LOGIN_WITH_VALIDATE_CODE_2FA_WORKSPACE_CARD]: ROUTES.LOGIN_WITH_VALIDATE_CODE_2FA_WORKSPACE_CARD,
+            LogInWithShortLivedToken: ROUTES.LOGIN_WITH_SHORTLIVED_AUTHTOKEN,
```
Comment (cead22) [inline src/libs/Navigation/linkingConfig.js:35]: Should we put `LogInWithShortLivedToken` in a const under SCREEN?

File: src/libs/actions/Session.js [hunk 2]
```diff
@@ -234,6 +237,29 @@ function signIn(password, twoFactorAuthCode) {
         });
 }
 
+function signInWithShortLivedToken(accountID, email, shortLivedToken, encryptedAuthToken, exitTo) {
+    Onyx.merge(ONYXKEYS.ACCOUNT, {...CONST.DEFAULT_ACCOUNT_DATA, loading: true});
+
+    createTemporaryLogin(shortLivedToken, encryptedAuthToken).then((response) => {
```
Comment (cead22) [inline src/libs/actions/Session.js:243]: Is createTemporaryLogin missing the email param? Maybe it's not really needed in `API.CreateLogin`, but NAB we don't have to fix that in this PR

File: src/pages/LogInWithShortLivedTokenPage.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    componentDidMount() {
+        const accountID = parseInt(lodashGet(this.props.route.params, 'accountID', ''), 10);
+        const email = lodashGet(this.props.route.params, 'email', '');
+        const shortLivedToken = lodashGet(this.props.route.params, 'shortLivedToken', '');
+
+        // exitTo is URI encoded because it could contain a variable number of slashes (i.e. "workspace/new" vs "workspace/<ID>/card")
+        const exitTo = decodeURIComponent(lodashGet(this.props.route.params, 'exitTo', ''));
+
+        // If the user is revisiting the component authenticated or they were already logged into the right account, we simply redirect them to the exitTo
+        if (this.props.session.authToken && email === this.props.session.email) {
```
Comment (marcaaron) [inline src/pages/LogInWithShortLivedTokenPage.js:59]: NAB, it's good that we are checking if the email is the same. But it doesn't look like we are handling the case where it's not. We might not need to worry about this just yet, but feels like bad things would happen if you tried to sign in one user while still logged in as another.

File: src/libs/actions/Session.js [hunk 4]
```diff
@@ -234,6 +237,27 @@ function signIn(password, twoFactorAuthCode) {
         });
 }
 
+function signInWithShortLivedToken(accountID, email, shortLivedToken) {
```
Comment (marcaaron) [inline src/libs/actions/Session.js:240]: Add some docs here?

**Review by marcaaron — COMMENTED** [review]
Leaving a few initial comments.

File: src/libs/actions/Session.js [hunk 5]
```diff
@@ -234,6 +237,27 @@ function signIn(password, twoFactorAuthCode) {
...
+    Onyx.merge(ONYXKEYS.ACCOUNT, {...CONST.DEFAULT_ACCOUNT_DATA, loading: true});
+
+    createTemporaryLogin(shortLivedToken).then((response) => {
+        Onyx.merge(ONYXKEYS.SESSION, {
+            authToken: shortLivedToken,
+            accountID,
+            email,
+        });
+        if (response.jsonCode === 200) {
+            getUserDetails();
```
Comment (marcaaron) [inline src/libs/actions/Session.js:250]: Oh hmm why are we doing this? It should happen when we set the `authToken` no? 

https://github.com/Expensify/App/blob/1901df375c91eef883f455caf8b5dadeddb9bf9f/src/libs/Navigation/AppNavigator/AuthScreens.js#L185-L186

File: src/libs/actions/Session.js (hunk 4 above)
Comment (cead22) [inline src/libs/actions/Session.js:240]: - Don't we need to call authenticate with the short lived token before creating a login?
- We're calling `signInWithShortLivedToken(shortLivedToken)` and not passing encrypted auth token, that means it'll be set to undefined in `setSuccessfulSignInData` right?

File: src/pages/LogInWithShortLivedTokenPage.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import PropTypes from 'prop-types';
+import {withOnyx} from 'react-native-onyx';
+import compose from '../libs/compose';
+import ONYXKEYS from '../ONYXKEYS';
+import {signInWithShortLivedToken} from '../libs/actions/Session';
+import FullScreenLoadingIndicator from '../components/FullscreenLoadingIndicator';
+import Navigation from '../libs/Navigation/Navigation';
+
+const propTypes = {
+    /** The accountID and validateCode are passed via the URL */
```
Comment (cead22) [inline src/pages/LogInWithShortLivedTokenPage.js:12]: Is this right or outdated?

File: src/pages/LogInWithShortLivedTokenPage.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            shortLivedToken: PropTypes.string,
+
+            // URL to exit to
+            exitTo: PropTypes.string,
+        }),
+    }),
+
+    /** The data about the current session */
+    session: PropTypes.shape({
+        /** The authToken for the current session */
```
Comment (cead22) [inline src/pages/LogInWithShortLivedTokenPage.js:35]: Some use `/**` and some `//`

File: src/SCREENS.js [hunk 8]
```diff
@@ -6,6 +6,7 @@ export default {
     HOME: 'Home',
     LOADING: 'Loading',
     REPORT: 'Report',
+    LOG_IN_WITH_SHORT_LIVED_TOKEN: 'LogInWithShortLivedToken',
```
Comment (Julesssss) [inline src/SCREENS.js:9]: Can we align `SHORTLIVED` and `SHORT_LIVED` to help make this searchable?

Comment (iwiznia) [reply src/SCREENS.js:9]: Oh true. Yes, I will once I figure out which PR we should be using.

**Comment by iwiznia** [comment]
This PR was replaced by https://github.com/Expensify/App/pull/5396/

**Pull request closed without merging**